src/Microsoft.ML.Core/Data/Repository.cs

```diff
@@ -256,7 +256,7 @@ protected void GetPath(out string pathEnt, out string pathTemp, string dir, stri
             string root = Path.GetFullPath(DirTemp ?? @"x:\dummy");
             string entityPath = Path.Combine(root, dir ?? "", name);
             entityPath = Path.GetFullPath(entityPath);
-            string tempPath = Path.Combine(root, PathMap.Count.ToString());
+            string tempPath = Path.Combine(root, Guid.NewGuid().ToString());
```

Comment:
Can you please help me understand the problem you are trying to fix?

Reply:
Currently, there's a race condition in creating temporary files, causing a naming collision. 

The paths are created as `C:\Users\jormont\AppData\Local\Temp\2\TLC_25553D77\0`, where:
1. `TLC_` is fixed
2. `25553D77` is a seeded random in hex (bad)
3. `0` is the count [0..N] of folders existing in the parent folder (also bad).

The prefix `TLC_` (1) is fine, though should be updated to the current name, perhaps to `ML_NET_`. Or `Microsoft_ML_`, though I'd worry about creating too long of paths under Windows ~260 chars, especially as a we move towards a GUID. Note users can have long usernames, plus the names of the files placed in this path.

First issue -- The seeded random number (2) is an issue since running parallelly (or rerunning) with the same seed creates the same "random" folder name. This part is a deterministic collision.

Second issue -- The race condition then occurs when counting how many folders (3) are in the parent folder. When multiple processes/threads check the contents of the folder, they all receive the same number.  The purposed fix replaces this count [0..N] with a GUID.

The need is to have unique folders for creating temporary files, otherwise they clobber each other. This causing a crash and has the possibility of silent failures, especially on macOS/Linux where you can delete and overwrite files with open handles.

Reply:
Pushing deeper into the code, the random number (2) is not seeded, but uses the system seed which is set from the current time. I expect the collision occurs due the remark in `System.Random` docs, "On most Windows systems, Random objects created within 15 milliseconds of one another are likely to have identical seed values." ([src](https://docs.microsoft.com/en-us/dotnet/api/system.random?view=netcore-3.1)). Hence another race condition in (2).

Hopefully the .NET GUIDs don't have the same 15ms race condition as the RNG.

Reply:
In the current code, it looks like we are using our own random number generation, which seems odd. 
It is okay to use GUIDs. But they seem a bit of overkill. 

[Path.GetRandomFileName](https://docs.microsoft.com/en-us/dotnet/api/system.io.path.getrandomfilename?view=netframework-4.8) seems to be intended for this kind of scenario. 
Would that work better?

Reply:
@harishsk: You're right, `Path.GetRandomFileName()` seems more purpose built than a GUID.

Code behind this function -- https://github.com/dotnet/runtime/blob/f57d49bb711fda8ac0bd231229d91919f4fca8e8/src/libraries/System.Private.CoreLib/src/System/IO/Path.cs#L249-L265

Reply:
@harishsk I had no idea that method existed. Thanks for letting me know! Updated to use it.